### PR TITLE
Deprecate several flags in rustdoc

### PR DESCRIPTION
**Pull request by steveklabnik**
Part of #44136 

cc @rust-lang/dev-tools @rust-lang/docs 

This is a very basic PR to start deprecating some flags; `rustdoc` doesn't really have fancy output options like `rustc` does, so I went with `eprintln!`. Happy to change it if people feel that's not appropriate.

Also, I have no idea if we can or should write tests here, so I didn't try. If someone feels strongly about it, then let's do it, but given that the only outcome here is a side effect...

### CODE REVIEW

File: src/librustdoc/lib.rs
Comment (Manishearth) [inline src/librustdoc/lib.rs]: Nl at eof

**Review by Manishearth — APPROVED** [review]
r=me. Probs should use the colorful one over eprintln

**Comment by steveklabnik** [comment]
> Probs should use the colorful one

What's "the colorful one"?

**Comment by Manishearth** [comment]
nvm, that's only available from within rustc and nontrivial to use without a `Session`

**Comment by frewsxcv** [comment]
@bors r-

https://travis-ci.org/rust-lang/rust/jobs/269657868

**Comment by alexcrichton** [comment]
Hm right now this just prints out "WARNING: this flag is considered deprecated", could it also print the flag as well?

**Comment by steveklabnik** [comment]
Great, so this works with my Cargo PR.

@alexcrichton , what's the process for updating Cargo these days? It looks like it depends on a released version, I'm not sure what the protocol is for that.

> Hm right now this just prints out "WARNING: this flag is considered deprecated", could it also print the flag as well?

Also happy to do that as well, alongside dropping the testing changes to the submodules.

**Comment by alexcrichton** [comment]
Oh you can just update the cargo submodule whenever on the master branch of rust, no worries!

**Comment by steveklabnik** [comment]
```text
> rustdoc +local --no-defaults
WARNING: the 'no-defaults' flag is considered deprecated
WARNING: please see https://github.com/rust-lang/rust/issues/44136
```

🤘

**Comment by steveklabnik** [comment]
@alexcrichton https://travis-ci.org/rust-lang/rust/jobs/270119195#L772

what's the right way to resolve this? change it to 0.23?

**Comment by alexcrichton** [comment]
Probably rebase on top of https://github.com/rust-lang/rust/pull/44154 and wait for that to land I think?

**Comment by steveklabnik** [comment]
Ah ha, okay, will do.

**Comment by carols10cents** [comment]
Looks like this is ready for rereview? @GuillaumeGomez @Manishearth @alexcrichton one of yinz want to r+ this?

**Comment by alexcrichton** [comment]
The `--passes` flag is being deprecated here with a mention of its only use being documenting private types, but this isn't providing any ability for those users to do something that doesn't generate warnings. I don't personally have an opinion one way or another, but wanted to double check if that's intended or not?

**Comment by steveklabnik** [comment]
> I don't personally have an opinion one way or another, but wanted to double check if that's intended or not?

We want to introduce a new flag that will just do this; given that it's effectively an alias, I wanted to land this first, and then land that afterward. If you wanted it to be done in the same PR, I'd be happy to do that as well.

(I am kind of interested in how many people use this functionality and if there's any other constraints that are missing before we end up adding another insta-stable flag.)

**Comment by alexcrichton** [comment]
Ok r=me either way, I'll let you decide if you'd like to handle that use case

**Comment by alexcrichton** [comment]
@steveklabnik any thoughts on this? You're alright landing with a warning that has no replacement?

**Comment by kennytm** [comment]
@steveklabnik Ping. Any update?

**Comment by steveklabnik** [comment]
Update is, I want to add the new flag before merging. I will do it soon.

**Comment by steveklabnik** [comment]
Okay so, I have to rebuild LLVM locally, and this pr isn't *done* yet, but since we're introducing the new flag, I'd like everyone to weigh in on it: `--document-private-items`. It's a bit long, but I couldn't come up with anything snappier.

Before this gets merged, it needs to be 1. tested 2. suggested when one of the deprecated flags is used.

@rfcbot fcp merge

**Comment by steveklabnik** [comment]
Okay so, I have to rebuild LLVM locally, and this pr isn't *done* yet, but since we're introducing the new flag, I'd like everyone to weigh in on it: `--document-private-items`. It's a bit long, but I couldn't come up with anything snappier.

Before this gets merged, it needs to be 1. tested 2. suggested when one of the deprecated flags is used.

@rfcbot fcp merge

**Comment by steveklabnik** [comment]
This should now be ready to merge; I've tested it locally and added an extra notice.

I've also created the correct label, let's see if rfcbot sees this:

@rfcbot fcp merge

**Comment by QuietMisdreavus** [comment]
[00:03:19] tidy error: /checkout/src/librustdoc/lib.rs:581: line longer than 100 chars
    [00:03:19] some tidy checks failed

Also, i think you would have both `T-dev-tools` and `T-dev-tools-rustdoc`? That way the core dev-tools team gets in and it's more than just me and imperio. That's all assuming rfcbot actually sees the call for FCP, which i don't think will work here?

**Comment by steveklabnik** [comment]
@QuietMisdreavus hehe already fixed ❤️ 

I've added the tag, so let's try it again. I know the libs team uses rfcbot in this repo, so it *should* work

@rfcbot fcp merge

**Comment by QuietMisdreavus** [comment]
I'm also remembering https://github.com/rust-lang/rust/pull/43949#issuecomment-328564813 where it didn't work for dev-tools in the past. `>_>`

**Comment by steveklabnik** [comment]
Okay well, sorry rfcbot i'm done fighting with you :)

ping the @rust-lang/dev-tools team for signoffs; here's some manual ticky boxes for you:

- [ ] @nrc 
- [ ] @michaelwoerister 
- [ ] @killercup 
- [ ] @QuietMisdreavus 
- [ ] @GuillaumeGomez 
- [ ] @steveklabnik 
- [ ] @jonathandturner 
- [ ] @japaric

Primary sign-off here is "`--document-private-items` as the name of the flag for rustdoc to document private items". I'm open to suggestions.

**Comment by killercup** [comment]
LGTM

@klabnikatron reviewed

**Comment by alexcrichton** [comment]
@steveklabnik loos like the last remaining box here is @nrc whom I believe is on parental leave right now, although I doubt he'd object to this :)

**Comment by michaelwoerister** [comment]
We looked at this during dev-tools triage. As @alexcrichton says, I think this is good to go, @steveklabnik!

**Comment by kennytm** [comment]
Ping again @steveklabnik 🙂. I think the teams' decisions are "it is good to go". The remaining item is to [get it tested](https://github.com/rust-lang/rust/pull/44138#issuecomment-331238218), right?

**Comment by steveklabnik** [comment]
Sorry for the delay here, will get on this tomorrow!

**Comment by steveklabnik** [comment]
Okay; we don't have a great way to test the output here, but on @QuietMisdreavus 's suggestion, I've changed a few of the rustdoc tests to use the new flag. That gives us about as much coverage as the old flags did.

**Comment by QuietMisdreavus** [comment]
\*drumroll\*

:shipit:

@bors r+

**Pull request merged**
